File: SquirrelSave/shared/gamification.ts
import { BUDGET, GAMIFICATION, type MascotMood } from "./config";

export function computeLevel(xp: number, xpPerLevel = GAMIFICATION.xpPerLevel): number {
  return Math.floor(xp / xpPerLevel) + 1;
}

export function xpToNextLevel(xp: number, xpPerLevel = GAMIFICATION.xpPerLevel): number {
  return xpPerLevel - (xp % xpPerLevel);
}

export function computeMascotMood(
  spendingPercent: number,
  streak: number,
  hasBudgetAlert = false
): MascotMood {
  if (hasBudgetAlert) return "worried";
  if (streak === 0 && spendingPercent === 0) return "sleeping";
  if (streak >= GAMIFICATION.streakCelebrationDays && spendingPercent < GAMIFICATION.spendingCelebratingMax) {
    return "celebrating";
  }
  if (spendingPercent >= GAMIFICATION.spendingAlertMin) return "alert";
  if (spendingPercent >= GAMIFICATION.spendingWorriedMin) return "worried";
  return "happy";
}

export function isBudgetAlert(spendingPercent: number): boolean {
  return spendingPercent >= BUDGET.alertThresholdPercent;
}

File: SquirrelSave/backend/app/gamification.py
def compute_level(xp: int, xp_per_level: int = 500) -> int:
    return xp // xp_per_level + 1


def xp_to_next_level(xp: int, xp_per_level: int = 500) -> int:
    return xp_per_level - (xp % xp_per_level)


def compute_mascot_mood(
    spending_percent: float,
    streak: int,
    has_budget_alert: bool = False,
) -> str:
    if has_budget_alert:
        return "worried"
    if streak == 0 and spending_percent == 0:
        return "sleeping"
    if streak >= 7 and spending_percent < 60:
        return "celebrating"
    if spending_percent >= 90:
        return "alert"
    if spending_percent >= 70:
        return "worried"
    return "happy"


def is_budget_alert(spending_percent: float, threshold: int = 80) -> bool:
    return spending_percent >= threshold
